near/omni-bridge/src/lib.rs

```diff
@@ -489,18 +496,89 @@ impl Contract {
         );
 
         let mut required_storage_balance =
-            self.add_transfer_message(transfer_message.clone(), storage_payer.clone());
+            self.required_balance_for_init_transfer_message(transfer_message.clone());
         required_storage_balance = required_storage_balance.saturating_add(
             NearToken::from_yoctonear(init_transfer_msg.native_token_fee.0),
         );
 
+        let message_storage_account_id =
+            Self::calculate_message_storage_account_id(&transfer_message, &signer_id);
+        if self.has_storage_balance(&message_storage_account_id, required_storage_balance) {
+            self.init_transfer_internal(transfer_message, message_storage_account_id, signer_id);
+        } else if self.has_storage_balance(&signer_id, required_storage_balance) {
+            self.init_transfer_internal(transfer_message, signer_id.clone(), signer_id);
+        } else {
+            let promise_index = env::promise_yield_create(
+                "init_transfer_resume",
+                json!({
+                    "transfer_message": transfer_message,
+                    "message_storage_account_id": message_storage_account_id,
+                    "storage_owner": signer_id,
+                })
+                .to_string()
+                .as_bytes(),
+                INIT_TRANSFER_RESUME_GAS,
+                GasWeight(0),
+                PROMISE_REGISTER_ID,
+            );
+
+            let yield_id: CryptoHash = env::read_register(PROMISE_REGISTER_ID)
+                .expect("ERR_READ_PROMISE_REGISTER")
+                .try_into()
+                .expect("ERR_READ_PROMISE_REGISTER");
+
+            self.init_transfer_promises
+                .insert(&message_storage_account_id, &yield_id);
+
+            env::promise_return(promise_index);
+        }
+    }
+
+    #[private]
+    #[allow(clippy::needless_pass_by_value)]
+    pub fn init_transfer_resume(
+        &mut self,
+        transfer_message: TransferMessage,
+        message_storage_account_id: AccountId,
+        storage_owner: AccountId,
+        #[callback_result] response: Result<(), PromiseError>,
+    ) {
+        self.init_transfer_promises
+            .remove(&message_storage_account_id);
+
+        match response {
+            Ok(()) => self.init_transfer_internal(
```

Comment:
if the `init_transfer_internal` panic then the promise will not be removed from the storage